nova_plugin/server.py

```diff
@@ -548,11 +548,21 @@ def attach_volume(nova_client, cinder_client, **kwargs):
         server_id,
         volume_id,
         device if device != 'auto' else None)
-    volume.wait_until_status(cinder_client=cinder_client,
-                             volume_id=volume_id,
-                             status=volume.VOLUME_STATUS_IN_USE)
-    if device == 'auto':
+    try:
+        vol, wait_succeeded = volume.wait_until_status(
+            cinder_client=cinder_client,
+            volume_id=volume_id,
+            status=volume.VOLUME_STATUS_IN_USE
+        )
+        if not wait_succeeded:
+            raise RecoverableError(
+                'Waiting for volume status {0} failed - detaching volume and '
+                'retrying..'.format(volume.VOLUME_STATUS_IN_USE))
+    except RecoverableError:
```

Comment:
couple of issues:
1) there's another API call below - "volume.get_attachment" - that too may fail, upon which a revert will also be needed. i guess whats needed is to pretty much wrap up everything beyond the first API call which actually modifies something with a try-except block.
2) the exception caught is only of RecoverableError - note that even if a different error (one which is not an instance of RecoverableError) gets raised, as long as its not a NonRecoverableError, the operation will get retried. This means that any exception should be caught, and only NonRecoverableErrors should be re-raised without reverting first.
2)